clang/include/clang/Sema/SemaOpenACC.h

```diff
@@ -0,0 +1,74 @@
+//===----- SemaOpenACC.h - Semantic Analysis for OpenACC constructs -------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+/// \file
+/// This file declares semantic analysis for OpenACC constructs and
+/// clauses.
+///
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CLANG_SEMA_SEMAOPENACC_H
+#define LLVM_CLANG_SEMA_SEMAOPENACC_H
+
+#include "clang/AST/DeclGroup.h"
+#include "clang/Basic/OpenACCKinds.h"
+#include "clang/Basic/SourceLocation.h"
+#include "clang/Sema/Ownership.h"
+
+namespace clang {
+
+class ASTContext;
+class DiagnosticEngine;
+class LangOptions;
+class Sema;
+
+class SemaOpenACC {
+public:
+  SemaOpenACC(Sema &S);
+
+  ASTContext &getASTContext() const;
+  DiagnosticsEngine &getDiagnostics() const;
+  const LangOptions &getLangOpts() const;
+
+  Sema &SemaRef;
```

Comment:
Would it make sense to have a base class to provide this functionality so that `SemaOpenACC`, `SemaOpenMP`, `SemaWhateverElse`, etc can all inherit the same interfaces? And perhaps these interfaces should be private rather than public?

Reply:
Base class is certainly an option, but I intentionally try to avoid over-generalization in this PR. I'd like us to have more experience applying what is already in this patch, to make more informed decision whether we want base class.

Reply:
I don't think it's over-generalization to go ahead and extract these basic accessors into a common base class.  Just make sure you keep it to these common accessors.

Reply:
Unlike other decisions this patch forces onto `Sema` implementation and users, this one is a choice of implementation that is not supposed to force changes in the codebase, at least not in a major way. So I don't think we should stretch the scope of this PR with something that can be decided later.

Reply:
I think this work needs to be done at some point, but I don't think it needs to be in this commit if you'd prefer to do it as a follow-up when you go to add another sema "sub" class. Then we can argue about things like "protected access?" and the like without distracting from the incremental progress in this PR.